nexus/src/external_api/device_auth.rs

```diff
@@ -0,0 +1,262 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Entrypoints for the OAuth 2.0 Device Authorization Grant flow.
+//!
+//! These are endpoints used by the API client per se (e.g., the CLI),
+//! *not* the user of that client (e.g., an Oxide rack operator). They
+//! are for requesting access tokens that will be managed and used by
+//! the client to make other API requests.
+
+use super::console_api::{get_login_url, serve_console_index};
+use super::views::{DeviceAccessTokenGrant, DeviceAuthResponse};
+use crate::context::OpContext;
+use crate::db::model::DeviceAccessToken;
+use crate::ServerContext;
+use dropshot::{
+    endpoint, HttpError, HttpResponseOk, Query, RequestContext, TypedBody,
+};
+use http::{header, Response, StatusCode};
+use hyper::Body;
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+use serde_urlencoded;
+use std::sync::Arc;
+use uuid::Uuid;
+
+// Token granting à la RFC 8628 (OAuth 2.0 Device Authorization Grant)
+
+/// OAuth 2.0 error responses use 400 (Bad Request) with specific `error`
+/// parameter values to indicate protocol errors (see RFC 6749 §5.2).
+/// This is different from Dropshot's error `message` parameter, so we
+/// need a custom response builder.
+fn build_oauth_response<T>(
+    status: StatusCode,
+    body: &T,
+) -> Result<Response<Body>, HttpError>
+where
+    T: ?Sized + Serialize,
+{
+    let body = serde_json::to_string(body)
+        .map_err(|e| HttpError::for_internal_error(e.to_string()))?;
+    Ok(Response::builder()
+        .status(status)
+        .header(header::CONTENT_TYPE, "application/json")
+        .body(body.into())?)
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize, JsonSchema)]
+pub struct DeviceAuthRequestParams {
+    pub client_id: Uuid,
+}
+
+/// Start an OAuth 2.0 Device Authorization Grant
+///
+/// This endpoint is designed to be accessed from an *unauthenticated*
+/// API client. It generates and records a `device_code` and `user_code`
+/// which must be verified and confirmed prior to a token being granted.
+#[endpoint {
+    method = POST,
+    path = "/device/auth",
+    content_type = "application/x-www-form-urlencoded",
+    tags = ["hidden"], // "token"
```

Comment:
This seems important for OAuth, but can we set `unpublished = true` since I don't think we expect other consumers to use this?